acme/environment_loops/open_spiel_environment_loop.py

```diff
@@ -0,0 +1,221 @@
+# limitations under the License.
+
+"""An OpenSpiel multi-agent/environment training loop."""
+
+import operator
+import time
+from typing import Optional, Sequence
+
+from acme import core
+from acme.utils import counting
+from acme.utils import loggers
+from acme.wrappers import open_spiel_wrapper
+import dm_env
+from dm_env import specs
+import numpy as np
+# pytype: disable=import-error
+import pyspiel
+# pytype: enable=import-error
+import tree
+
+
+class OpenSpielEnvironmentLoop(core.Worker):
+  """An OpenSpiel RL environment loop.
+
+  This takes `Environment` and list of `Actor` instances and coordinates their
+  interaction. Agents are updated if `should_update=True`. This can be used as:
+
+    loop = EnvironmentLoop(environment, actors)
+    loop.run(num_episodes)
+
+  A `Counter` instance can optionally be given in order to maintain counts
+  between different Acme components. If not given a local Counter will be
+  created to maintain counts between calls to the `run` method.
+
+  A `Logger` instance can also be passed in order to control the output of the
+  loop. If not given a platform-specific default logger will be used as defined
+  by utils.loggers.make_default_logger. A string `label` can be passed to easily
+  change the label associated with the default logger; this is ignored if a
+  `Logger` instance is given.
+  """
+
+  def __init__(
+      self,
+      environment: open_spiel_wrapper.OpenSpielWrapper,
+      actors: Sequence[core.Actor],
+      counter: counting.Counter = None,
+      logger: loggers.Logger = None,
+      should_update: bool = True,
+      label: str = 'open_spiel_environment_loop',
+  ):
+    # Internalize agent and environment.
+    self._environment = environment
+    self._actors = actors
+    self._counter = counter or counting.Counter()
+    self._logger = logger or loggers.make_default_logger(label)
+    self._should_update = should_update
+
+    # Track information necessary to coordinate updates among multiple actors.
+    self._observed_first = [False] * len(self._actors)
+    self._prev_actions = [pyspiel.INVALID_ACTION] * len(self._actors)
+
+  def _send_observation(self, timestep: dm_env.TimeStep, player: int):
+    # If terminal all actors must update
+    if player == pyspiel.PlayerId.TERMINAL:
+      for player_id in range(len(self._actors)):
+        # Note: we must account for situations where the first observation
+        # is a terminal state, e.g. if an opponent folds in poker before we get
+        # to act.
+        if self._observed_first[player_id]:
+          player_timestep = self._get_player_timestep(timestep, player_id)
+          self._actors[player_id].observe(self._prev_actions[player_id],
+                                          player_timestep)
+          if self._should_update:
+            self._actors[player_id].update()
+      self._observed_first = [False] * len(self._actors)
+      self._prev_actions = [pyspiel.INVALID_ACTION] * len(self._actors)
```

Comment:
Maybe initialize these two once at the beginning of run_episode? That would eliminate code duplication.